diff --git a/arch/arm/kernel/vmlinux.lds.S b/arch/arm/kernel/vmlinux.lds.S
index 8af654bd1..1e9ff41d6 100644
--- a/arch/arm/kernel/vmlinux.lds.S
+++ b/arch/arm/kernel/vmlinux.lds.S
@@ -131,6 +131,14 @@ SECTIONS
 		__pv_table_end = .;
 	}
 
+#ifdef CONFIG_DRIVERS_HDF
+	.init.hdf_table : {
+		_hdf_drivers_start = .;
+		*(.hdf.driver)
+		_hdf_drivers_end = .;
+	}
+#endif
+
 	INIT_DATA_SECTION(16)
 
 	.exit.data : {
diff --git a/drivers/Kconfig b/drivers/Kconfig
index dcecc9f6e..8eddfc2cd 100644
--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -234,5 +234,7 @@ source "drivers/interconnect/Kconfig"
 
 source "drivers/counter/Kconfig"
 
+source "drivers/hdf/khdf/Kconfig"
+
 source "drivers/most/Kconfig"
 endmenu
diff --git a/drivers/Makefile b/drivers/Makefile
index 576228037..025b92b1f 100644
--- a/drivers/Makefile
+++ b/drivers/Makefile
@@ -188,4 +188,5 @@ obj-$(CONFIG_SIOX)		+= siox/
 obj-$(CONFIG_GNSS)		+= gnss/
 obj-$(CONFIG_INTERCONNECT)	+= interconnect/
 obj-$(CONFIG_COUNTER)		+= counter/
+obj-$(CONFIG_DRIVERS_HDF)	+= hdf/
 obj-$(CONFIG_MOST)		+= most/
diff --git a/drivers/hdf/Makefile b/drivers/hdf/Makefile
new file mode 100644
index 000000000..5c5e1911c
--- /dev/null
+++ b/drivers/hdf/Makefile
@@ -0,0 +1,2 @@
+export PROJECT_ROOT := ../../../../../
+obj-$(CONFIG_DRIVERS_HDF) += khdf/
-- 
2.25.1

